tests/kernel/smp/src/main.c

```diff
@@ -37,45 +52,354 @@ void t2_fn(void *a, void *b, void *c)
-			ok = 0;
-			break;
-		}
+		zassert_true(t2_count > i, "Another thread isn't"
+			     " scheduled to run on other core");
+	}
+	k_thread_abort(tid);
+}
+
+static void smp_thread_entry(void *p1, void *p2, void *p3)
+{
+	int tnum = (int)p1;
+
+	TC_PRINT("Thread started\n");
+	k_busy_wait(THREAD_DELAY);
+
+	tdata[tnum].executed = 1;
+	TC_PRINT("Thread execution complete\n");
+}
+
+/**
+ * @brief Verify round robin on one core with coop thread
+ *
+ * @details The cooperative test creates THREAD_NUM preemptive
+ * threads of same priority and checks the round robin behavior.
+ * the cooperative thread will be blocking one core hence
+ * preemptive threads should run on another core in round robin.
+ */
+static void test_smp_single_core_coop_rr_threads(void)
+{
+	int i;
+
+	k_sched_time_slice_set(100, 0);
+
+	for (i = 0; i < THREADS_NUM; i++) {
+		tdata[i].executed = 0;
+		tdata[i].tid = k_thread_create(&tthread[i], tstack[i],
+					       STACK_SIZE, smp_thread_entry,
+					       (void *)i, NULL, NULL,
+					       K_PRIO_PREEMPT(
+						       K_LOWEST_THREAD_PRIO - 1),
+					       0, 0);
+	}
+
+	k_busy_wait(THREAD_DELAY * THREADS_NUM + 10);
+
+	for (i = 0; i < THREADS_NUM; i++) {
+		k_thread_abort(tdata[i].tid);
+	}
+	/* Test point: All THREADS_NUM spawned threads must have
+	 * executed in another core
+	 */
+	for (i = 0; i < THREADS_NUM; i++) {
+		zassert_true(tdata[i].executed == 1, "Threads didn't get"
+			     " scheduled on another core");
+	}
+}
+
+/**
+ * @brief Verify round robin on one core with preempt thread
+ *
+ * @details The preemptive thread of higher priority creates
+ * lower priority threads. So higher priority thread should occupy
+ * one core and lower priority threads should run in round
+ * robin.
+ */
+static void test_smp_single_core_preemp_rr_threads(void)
+{
+	int i;
+
+	k_thread_priority_set(k_current_get(),
+			      K_PRIO_PREEMPT(K_HIGHEST_THREAD_PRIO));
+
+	k_sched_time_slice_set(100, 0);
+
+	for (i = 0; i < THREADS_NUM; i++) {
+		tdata[i].executed = 0;
+		tdata[i].tid = k_thread_create(&tthread[i], tstack[i],
+					       STACK_SIZE, smp_thread_entry,
+					       (void *)i, NULL, NULL,
+					       K_PRIO_PREEMPT(
+						       K_LOWEST_THREAD_PRIO - 1),
+					       0, 0);
+	}
+
+	k_busy_wait(THREAD_DELAY * THREADS_NUM + 10);
+
+	for (i = 0; i < THREADS_NUM; i++) {
+		k_thread_abort(tdata[i].tid);
+	}
+
+	/* Test point: All THREADS_NUM spawned threads must have
+	 * executed in another core
+	 */
+	for (i = 0; i < THREADS_NUM; i++) {
+		zassert_true(tdata[i].executed == 1, "Threads didn't get"
+			     " scheduled on another core");
+	}
+}
+
+/**
+ * @brief Verify round robin on multi core with coop thread
+ *
+ * @details A cooperative thread creates THREAD_NUM preemptive
+ * threads of equal priority. After creation, cooperative thread
+ * goes for sleep till all preemptive threads finish execution.
+ * It checks if all threads are executed and returns test execution
+ * results
+ */
```

Comment:
Not sure what you're testing here.  On SMP, those threads will all be schedulable as soon as k_thread_create() returns, so this isn't actually testing anything different than the other threads.  What's the case you're trying to exercise?